core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java

```diff
@@ -373,26 +265,28 @@ private void updateWeightAndStep(Set<EndpointAndStep> endpointAndSteps) {
                 final EndpointAndStep endpointAndStep = i.next();
                 final int step = endpointAndStep.incrementAndGetStep();
                 final Endpoint endpoint = endpointAndStep.endpoint();
-                if (step == totalSteps) {
+                if (step >= totalSteps) {
                     endpointsFinishedRampingUp.add(endpoint);
                     i.remove();
                 } else {
-                    final int calculated =
-                            weightTransition.compute(endpoint, step, totalSteps);
-                    final int currentWeight = Ints.constrainToRange(calculated, 0, endpoint.weight());
+                    final int currentWeight = computeWeight(endpoint, step);
                     endpointAndStep.currentWeight(currentWeight);
                 }
             }
         }
 
         private void close() {
-            EndpointsRampingUpEntry entry;
-            while ((entry = endpointsRampingUp.poll()) != null) {
-                entry.scheduledFuture.cancel(true);
-            }
+            rampingUpWindowsMap.values().forEach(e -> e.scheduledFuture.cancel(true));
         }
     }
 
+    private static int numStep(long rampingUpIntervalNanos, Ticker ticker, long createTimestamp) {
+        final long timePassed = ticker.read() - createTimestamp;
+        final int step = Ints.saturatedCast(timePassed / rampingUpIntervalNanos);
+        // there's no point in having an endpoint at step 0 (no weight), so we increment by 1
+        return step + 1;
```

Comment:
Question) Couldn't `step` be `Integer.MAX_VALUE` if the endpoints are updated after a long time since the last update?